api/cryptocom/client.go

```diff
@@ -0,0 +1,259 @@
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+const (
+	userEndpoint   = "/v2/user"
+	marketEndpoint = "/v2/market"
+)
+
+type LogFunc func(format string, args ...interface{})
+
+func defaultLogFunc(format string, args ...interface{}) {
+	log.Printf(format, args...)
+}
+
+type Transport interface {
+	ReadMessage() (int, []byte, error)
+	WriteMessage(int, []byte) error
+	Close() error
+}
+
+func (c *Connection) Type() string {
+	if c.IsPrivate {
+		return "private"
+	}
+
+	return "public"
+}
+
+type HTTPClient interface {
+	Post(url, contentType string, body io.Reader) (resp *http.Response, err error)
+	Get(url string) (resp *http.Response, err error)
+}
+
+type Connection struct {
+	Endpoint  string
+	IsPrivate bool
+	Transport
+	sync.Mutex
+}
+
+type Client struct {
+	publicConn  Connection
+	privateConn Connection
+	wsRootURL   string
+	restRootURL string
+	key         string
+	secret      string
+	privateSubs []string
+	publicSubs  []string
+	httpClient  HTTPClient
+	outbox      chan Response
+	LogFunc     LogFunc
+	wg          sync.WaitGroup
+}
+
+// New returns a pointer of Client struct
+func New(wsRootURL, restRootURL, key, secret string) *Client {
+	return &Client{
+		key:         key,
+		secret:      secret,
+		wsRootURL:   wsRootURL,
+		restRootURL: restRootURL,
+		outbox:      make(chan Response),
+		privateSubs: make([]string, 0),
+		publicSubs:  make([]string, 0),
+		httpClient:  &http.Client{},
+		LogFunc:     defaultLogFunc,
+	}
+}
+
+// Connect instansiate WS Connections
+func (c *Client) Connect() error {
+	publicWsEndpoint := c.wsRootURL + marketEndpoint
+	privateWsEndpoint := c.wsRootURL + userEndpoint
+
+	cnx, err := c.createConnection(publicWsEndpoint, false)
+	if err != nil {
+		return err
+	}
+	c.publicConn = cnx
+
+	cnx, err = c.createConnection(privateWsEndpoint, true)
+	if err != nil {
+		return err
+	}
+	c.privateConn = cnx
+
+	time.Sleep(3 * time.Second) // Cryptocom requires this sleep.
+	c.authenticate()
+
+	return nil
+}
+
+func (c *Client) Listen() <-chan Response {
+	go c.readConnection(c.publicConn)
+	go c.readConnection(c.privateConn)
+	return c.outbox
+}
+
+func (c *Client) Shutdown() {
+	c.privateConn.Close()
+	c.publicConn.Close()
+	c.wg.Wait()
+	close(c.outbox)
+}
+
+func (c *Client) createConnection(endpoint string, isPrivate bool) (Connection, error) {
+	conn, _, err := websocket.DefaultDialer.Dial(endpoint, http.Header{})
+	if err != nil {
+		return Connection{}, err
+	}
+
+	return Connection{Endpoint: endpoint, IsPrivate: isPrivate, Transport: conn}, nil
+}
+
+func (c *Client) readConnection(cnx Connection) {
+	defer c.wg.Done()
+	c.wg.Add(1)
+
+	c.LogFunc("Start listening connection ... %s", cnx.Endpoint)
+	for {
+		_, m, err := cnx.ReadMessage()
+		if err != nil {
+			c.LogFunc("error on read message in %s cnx\nError message: %s\n", cnx.Type(), err.Error())
+			if isClosedCnxError(err) {
```

Comment:
here better you check if terminating == true and only return on this case.
This way we make sure we break the reconnection loop only if we explicitly want to shutdown the client.